dev/sensor/bh1750/export/dev/sensor/bh1750.hpp

```diff
@@ -0,0 +1,239 @@
+    //! I2C slave address when ADDR pin is pulled up to VCC
+    static constexpr uint8_t i2s_address_high = 0xB8;
+
+    //! Maximum measurement time for low resolution mode according to RM
+    static constexpr uint16_t measurement_time_ms_low = 25;
+    //! Maximum measurement time for high and high2 resolution modes according to RM
+    static constexpr uint16_t measurement_time_ms_high = 200;
+
+    //! Current resolution mode
+    static bh1750_cfg::resolution resolution;
+
+    //! List of commands, supported by sensor
+    enum {
+        CMD_POWER_DOWN = 0x0,
+        CMD_POWER_UP = 0x1,
+        CMD_RESET = 0x7,
+        CMD_CONTINUOUSLY_HR_MODE = 0x10,
+        CMD_CONTINUOUSLY_HR_MODE2 = 0x11,
+        CMD_CONTINUOUSLY_LR_MODE = 0x13,
+        CMD_ONE_TIME_HR_MODE = 0x20,
+        CMD_ONE_TIME_HR_MODE2 = 0x21,
+        CMD_ONE_TIME_LR_MODE = 0x23,
+    };
+};
+
+template <class i2c_dev, bh1750_cfg::i2c_address address>
+bh1750_cfg::resolution bh1750<i2c_dev, address>::resolution = bh1750_cfg::resolution::high;
+
+template <class i2c_dev, bh1750_cfg::i2c_address address>
+ecl::err bh1750<i2c_dev, address>::init()
+{
+    return i2c_dev::init();
+}
+
+template <class i2c_dev, bh1750_cfg::i2c_address address>
+constexpr uint8_t bh1750<i2c_dev, address>::pick_i2c_slave_address()
+{
+    if (address == bh1750_cfg::i2c_address::low) {
+        return i2s_address_low;
+    } else {
+        return i2s_address_high;
+    }
+}
+
+template <class i2c_dev, bh1750_cfg::i2c_address address>
+void bh1750<i2c_dev, address>::set_resolution(bh1750_cfg::resolution resolution)
+{
+    bh1750<i2c_dev, address>::resolution = resolution;
+}
+
+template <class i2c_dev, bh1750_cfg::i2c_address address>
+bh1750_cfg::resolution bh1750<i2c_dev, address>::get_resolution()
+{
+    return resolution;
+}
+
+template <class i2c_dev, bh1750_cfg::i2c_address address>
+ecl::err bh1750<i2c_dev, address>::get_illuminance_sample(uint16_t &sample)
+{
+    uint8_t i2c_addr = pick_i2c_slave_address();
+
+    uint8_t cmd = CMD_ONE_TIME_HR_MODE;
+    uint16_t delay = measurement_time_ms_high;
+
+    switch (resolution) {
+    case bh1750_cfg::resolution::high:
+        cmd = CMD_ONE_TIME_HR_MODE;
+        delay = measurement_time_ms_high;
+        break;
+    case bh1750_cfg::resolution::high2:
+        cmd = CMD_ONE_TIME_HR_MODE2;
+        delay = measurement_time_ms_high;
+        break;
+    case bh1750_cfg::resolution::low:
+        cmd = CMD_ONE_TIME_LR_MODE;
+        delay = measurement_time_ms_low;
+        break;
+    default:
+        break;
+    }
+
+    i2c_dev::platform_handle().set_slave_addr(i2c_addr);
+
+    i2c_dev::lock();
+    err rc = i2c_dev::set_buffers(&cmd, nullptr, 1);
+    if (rc == err::ok) {
+        rc = i2c_dev::xfer();
+    }
+    i2c_dev::unlock();
+
+    if (rc != err::ok) {
+        return rc;
+    }
+
+    ecl::os::this_thread::sleep_for(delay);
+
+    uint8_t data[2] = {};
+
+    i2c_dev::lock();
```

Comment:
At this moment, someone else could already change slave address with `set_slave_addr()` call. Either set it again (which is preferred, I guess) or expand critical section (it has negative consequences of increased latency of the I2C response for other clients)